File: Lunara_backend/server/src/models/LegalDocument.ts
import { Model, DataTypes, Optional } from 'sequelize';
import sequelize from '../config/database';

export enum LegalDocumentType {
    TERMS_OF_SERVICE = 'terms_of_service',
    PRIVACY_POLICY = 'privacy_policy',
    REFUND_POLICY = 'refund_policy',
    OTHER = 'other',
}

export interface LegalDocumentAttributes {
    id: string;
    title: string;
    type: LegalDocumentType;
    content: string;
    version: string;
    isActive: boolean;
    effectiveDate: Date;
    createdAt?: Date;
    updatedAt?: Date;
}

export interface LegalDocumentCreationAttributes
    extends Optional<LegalDocumentAttributes, 'id' | 'isActive' | 'createdAt' | 'updatedAt'> {}

class LegalDocument
    extends Model<LegalDocumentAttributes, LegalDocumentCreationAttributes>
    implements LegalDocumentAttributes {
    public id!: string;
    public title!: string;
    public type!: LegalDocumentType;
    public content!: string;
    public version!: string;
    public isActive!: boolean;
    public effectiveDate!: Date;
    public readonly createdAt!: Date;
    public readonly updatedAt!: Date;
}

LegalDocument.init(
    {
        id: {
            type: DataTypes.UUID,
            defaultValue: DataTypes.UUIDV4,
            primaryKey: true,
        },
        title: {
            type: DataTypes.STRING(255),
            allowNull: false,
            validate: {
                notEmpty: { msg: 'Title is required' },
                len: { args: [2, 255], msg: 'Title must be between 2 and 255 characters' },
            },
        },
        type: {
            type: DataTypes.STRING(50),
            allowNull: false,
            validate: {
                isIn: {
                    args: [Object.values(LegalDocumentType)],
                    msg: `Type must be one of: ${Object.values(LegalDocumentType).join(', ')}`,
                },
            },
        },
        content: {
            type: DataTypes.TEXT,
            allowNull: false,
            validate: {
                notEmpty: { msg: 'Content is required' },
            },
        },
        version: {
            type: DataTypes.STRING(20),
            allowNull: false,
            validate: {
                notEmpty: { msg: 'Version is required' },
            },
        },
        isActive: {
            type: DataTypes.BOOLEAN,
            defaultValue: true,
            field: 'is_active',
        },
        effectiveDate: {
            type: DataTypes.DATEONLY,
            allowNull: false,
            field: 'effective_date',
            validate: {
                isDate: { msg: 'Must be a valid date', args: true },
            },
        },
    },
    {
        sequelize,
        tableName: 'legal_documents',
        underscored: true,
        timestamps: true,
        indexes: [
            { fields: ['type'] },
            { fields: ['is_active'] },
            { fields: ['effective_date'] },
        ],
    }
);

export default LegalDocument;

File: lib/models/legal_document.dart
class LegalDocument {
  final String id;
  final String title;
  final String type;
  final String content;
  final String version;
  final bool isActive;
  final DateTime? effectiveDate;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  LegalDocument({
    required this.id,
    required this.title,
    required this.type,
    required this.content,
    required this.version,
    required this.isActive,
    this.effectiveDate,
    this.createdAt,
    this.updatedAt,
  });

  factory LegalDocument.fromJson(Map<String, dynamic> json) {
    return LegalDocument(
      id: json['id']?.toString() ?? '',
      title: json['title'] ?? '',
      type: json['type'] ?? '',
      content: json['content'] ?? '',
      version: json['version']?.toString() ?? '',
      isActive: json['isActive'] ?? json['is_active'] ?? false,
      effectiveDate: json['effectiveDate'] != null 
          ? DateTime.tryParse(json['effectiveDate'].toString()) 
          : (json['effective_date'] != null ? DateTime.tryParse(json['effective_date'].toString()) : null),
      createdAt: json['createdAt'] != null 
          ? DateTime.tryParse(json['createdAt'].toString()) 
          : (json['created_at'] != null ? DateTime.tryParse(json['created_at'].toString()) : null),
      updatedAt: json['updatedAt'] != null 
          ? DateTime.tryParse(json['updatedAt'].toString()) 
          : (json['updated_at'] != null ? DateTime.tryParse(json['updated_at'].toString()) : null),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'title': title,
      'type': type,
      'content': content,
      'version': version,
      'isActive': isActive,
      'effectiveDate': effectiveDate?.toIso8601String(),
      'createdAt': createdAt?.toIso8601String(),
      'updatedAt': updatedAt?.toIso8601String(),
    };
  }
}
